lib/utils/reactivation_authenticator.dart

```diff
@@ -0,0 +1,92 @@
+import 'dart:async';
+
+import 'package:chopper/chopper.dart';
+import 'package:coffeecard/cubits/authentication/authentication_cubit.dart';
+import 'package:coffeecard/data/repositories/v1/account_repository.dart';
+import 'package:coffeecard/data/storage/secure_storage.dart';
+import 'package:coffeecard/service_locator.dart';
+
+class ReactivationAuthenticator extends Authenticator {
+  final SecureStorage secureStorage;
+  DateTime? tokenRefreshedAt;
+  final Duration debounce = const Duration(seconds: 10);
+  int _retryCount = 0;
+  final int _retryLimit = 1;
+
+  ReactivationAuthenticator(this.secureStorage);
+
+  bool _canRefreshToken() {
+    if (tokenRefreshedAt == null) {
+      return true;
+    } else {
+      return tokenRefreshedAt!.difference(DateTime.now()) < debounce;
+    }
+  }
+
+  Future<void> _evict() async {
+    final authenticationCubit = sl.get<AuthenticationCubit>();
+    await authenticationCubit.unauthenticated();
+  }
+
+  @override
+  FutureOr<Request?> authenticate(
+    Request request,
+    Response response, [
+    Request? originalRequest,
+  ]) async {
+    if (response.statusCode == 401) {
+      // sign the user out
+      if (_retryCount > _retryLimit) {
+        _evict();
+        return null;
+      }
+
+      // avoid refreshing the token multiple times if requests happen at the same time
+      if (!_canRefreshToken()) {
+        return null;
+      }
+
+      return await refreshToken(request, response);
+    }
+    return null;
+  }
+
+  Future<Request?> refreshToken(
+    Request request,
+    Response response,
+  ) async {
+    _retryCount++;
+
+    final accountRepository = sl.get<AccountRepository>();
```

Comment:
The service locator should be moved to the repository, else the code is non-testable

Reply:
i assume you mean that the ReactivationAuthenticator should accept the serviceLocator as parameter and should then retrieve the AccountRepository and AuthenticationCubit from it?

Reply:
Yes that's probably the only solution as we are not able to depend on the account repository with our current service locater 👍🏻